design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,398 @@
+
+## Guiding Principles
+
+Our users are the stakeholders on how we make decisions. There are two types of users for managed resources:
+* Input Users: Users who configure/create/manage the resource. Crossplane Claim Controller, Cluster Admin are a few
+  examples of this type. We want our `Spec` to be sole place for them to configure everything about the resource.
+* Output Users: Users who refer to/query/use/extract information from the managed resource. We want them to be able to
+  perform all read-only operations on the given resource as if they have the cloud provider's API available. This can
+  include configurations and metadata information about the resource. We also want them to access only `Status`
+  sub-resource of the managed resource because this is where the most up-to-date information about the resource is
+  stored and we are able to limit RBAC of these users to access only `Status`.
+  
+### Naming Conventions
+
+There are two naming decisions here to make:
+* Managed resource CR name
+* Identifier string of the external resource in the cloud provider.
+
+_Name_ in this context means the identifier string for the resource in its environment. CR name is the identifier
+for resource in Kubernetes environment. But for external resources, this can be either `id`, `uid` or `name` properties
+of the resource. The one we want to name is the one that is used when _referring to_ that resource in the same
+environment. For example, in AWS, VPCs do not have a name but `VpcID` properties that is used for identification.
+
+#### Custom Resource Instance Name
+
+The name of the managed resource CR, which lives in our control cluster. The claim controller who creates the managed
+resource gives its name or some other entity might create the managed resource, like user themselves. So, do not assume
+that all the managed resources will have the same naming format anywhere.
+
+Crossplane lives on Kubernetes and both the user and admin interacts with it through Kubernetes APIs. For better UX
+we want to mirror what Kubernetes uses for identification of a resource as much as possible, which is name and namespace
+pair. So, the claim controller should use these two entities of the Claim that it's processing when it decides on a name.
+However, the resources in the cloud provider are in the same environment, either per project (GCP) or same account (AWS).
+In order to provide a unique name that incorporates name and namespace, claim controller should make use of `GenerateName`
+feature of resource creation. The claim controller should assign `namespace-name` to `GenerateName` field of the `ObjectMeta`
+of the managed resource and Kubernetes will automatically append `-<5 characters random string>` to that when the creation
+is completed. For example, the following will be created by the claim controller:
+```yaml
+apiVersion: database.gcp.crossplane.io/v1alpha2
+kind: CloudsqlInstance
+metadata:
+  generateName: mydatabase-myappnamespace
```

Comment:
This looks like `name-namespace`, not `namespace-name` as you suggest earlier. I prefer `namespace-name`, because it leads with the least specific part of the identifier.